core/src/main/scala/cats/data/NonEmptySet.scala

```diff
@@ -0,0 +1,218 @@
+/*
+ * Copyright (c) 2018 Luka Jacobowitz
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cats
+package data
+
+import cats.instances.sortedSet._
+import cats.kernel._
+import cats.{Always, Eq, Eval, Foldable, Later, Now, Reducible, SemigroupK, Show}
+
+import scala.collection.immutable._
+
+final class NonEmptySet[A] private (val set: SortedSet[A]) extends AnyVal {
```

Comment:
`AnyVal` boxing can lead to a lot of reboxing when this is used in a generic context, which it would be in cats code since any of the typeclasses use it in generic context. I think we don't actually want `AnyVal`.

Reply:
Good point! Man, we really need opaque types :(